### ISSUE TITLE
support for taskgroups and cancelscopes in fixtures

### ISSUE THREAD

**Issue by graingert** [agronholm/anyio#345]
Currently documented as not supported
https://github.com/anyio/pull/76#issue-323360850

But pytest-trio has recently gained support for these https://pytest-trio.readthedocs.io/en/stable/reference.html#an-important-note-about-yield-fixtures

This could be done with `@anyio.pytest_fixture`

It would also be useful to expose this machinery for other similar tasks like asgi lifespan composition

**Comment by graingert** [comment]
See also how it could be used in FastAPI/Starlette https://github.com/tiangolo/fastapi/pull/3516#issuecomment-884485876

**Comment by mgrrx** [comment]
It would be really cool to have this feature to improve readability of tests. Somehow the first link https://github.com/anyio/pull/76#issue-323360850 does not work for me and I also couldn't find this limitation in the docs.

I experienced this issue myself and while debugging I came up with this minimal example here. I was about to file a ticket but then saw this ticket. Maybe it turns out to be useful.

```
import anyio
import pytest


@pytest.fixture
def anyio_backend():
    return "asyncio"


@pytest.fixture(name="server")
async def start_server(anyio_backend):
    async with anyio.create_task_group() as tg:
        yield 1


async def test_somethig(server):
    assert server == 1
```

which raises

```RuntimeError: Attempted to exit cancel scope in a different task than it was entered in```

(anyio 3.5.0)

**Comment by agronholm** [comment]
I know this would be really nice to have, and I've looked into implementing it, but the only way to implement it properly would be to commandeer the entire test session, and pytest does not really provide a way to do that. While it is true that pytest-trio has this capability, it is at the cost of async fixtures being limited to the function scope, plus horrendous added complexity. I've contemplated using a worker thread for running async fixtures on a background event loop, but receiving signals in tests would become a problem, as would potentially running subprocesses on older Linux kernels.
